### PR TITLE
Version 1.5

### PR DESCRIPTION
**Pull request by GeorgeGkas**
We update the software to version 1.5
---

### New simplified macros
Now we provide a cleaner option to generate errors without the need to provide the flag parameter. 

We defined macros based on the warning flags.

- `slog_none()`
- `slog_live()`
- `slog_info()`
- `slog_warn()`
- `slog_debug()`
- `slog_error()`
- `slog_fatal()`
- `slog_panic()`

Each macro take the following parameters:

 1. The log level.
 2. A formated string. Format tags prototype follows the same rules as the C standard library function `printf()`.
 3. Additional arguments. There should be the same number of these arguments as the number of `%-tags` that expect a value.

Bellow we provide an example that logs a debug message:

```c
slog_debug(0, "The %s contains between %d and %d billion stars and at least %d billion planets.  ", "Milky Way", 200, 400, 100);
```

In addition, we added the option to print the corresponding file name and line number where a slog macro was called. This rule follows the macros which relate to a critical flag, and shown bellow:

- `slog_warn()`
- `slog_error()`
- `slog_fatal()`
- `slog_panic()`

Basic example:
```c
/* Log and print something fatal. */
slog_fatal(0, "Fatal message. We fell into dead zone.");
```
With expected output to be:

    2017.01.22-19:03:17.03 - [FATAL] <example.c:71> -- Fatal message. We fell into dead zone.

### Updated Readme
We further updated the readme file. Now we provide better examples and explanations. Please refer to `Readme.md` file to read more.

### Updated Examples
We included more examples.

Please review this version. Thank you all.

### CODE REVIEW

File: src/slog.h [hunk 1]
```diff
@@ -27,44 +28,89 @@
...
+
+
+/* If compiled on DARWIN/Apple platforms. */
 #ifdef DARWIN
-#define    CLOCK_REALTIME    0x2d4e1588
-#define    CLOCK_MONOTONIC   0x0
-#endif //DARWIN
-
-/* Loging flags */
-#define SLOG_NONE   0
```
Comment (Burnett01) [inline src/slog.h:48]: I feel we should additionally keep these flags, in order not to break backward-compatibility. I've seen some closed-source projects that use these flags.

Proposed change:
```c
/* Loging flags. */
#define SLOG_FNONE   0
#define SLOG_FLIVE   1
#define SLOG_FINFO   2
#define SLOG_FWARN   3
#define SLOG_FDEBUG  4
#define SLOG_FERROR  5
#define SLOG_FFATAL  6
#define SLOG_FPANIC  7

/* Compat. Logging flags. */
#define SLOG_NONE   0
#define SLOG_LIVE   1
#define SLOG_INFO   2
#define SLOG_WARN   3
#define SLOG_DEBUG  4
#define SLOG_ERROR  5
#define SLOG_FATAL  6
#define SLOG_PANIC  7
````

File: README.md [hunk 2]
```diff
@@ -42,50 +49,100 @@ PRETTYLOG 0
...
+**Note:** `LOGFILELEVEL` and `LOGLEVEL` are completely independent of each other.
 
-LOGFILELEVEL and LOGLEVEL are completly independent of each other.
+**`LOGLEVEL` is passed as third argument in `slog_init()`.*   
+**`LOGFILELEVEL` is passed as fourth argument in `slog_init()`.*
 
 ### Logging flags
 Slog has its logging flags to print something with status code.
 
-- SLOG_LIVE
```
Comment (Burnett01) [inline README.md:58]: Let's additionally keep up with legacy flags. Please check my comments above.

Maybe add something along:

### Logging flags
 Slog has its logging flags to print something with status code.

- `SLOG_FLIVE`
- `SLOG_FINFO`
- `SLOG_FWARN`
- `SLOG_FDEBUG`
- `SLOG_FERROR`
- `SLOG_FFATAL`
- `SLOG_FPANIC`
- `SLOG_FNONE`

*Please note:* As of version 1.5 the flags have changed. In order to keep backward-compatibility, the following legacy flags can be used as usual:

- `SLOG_LIVE`
- `SLOG_INFO`
- `SLOG_WARN`
- `SLOG_DEBUG`
- `SLOG_ERROR`
- `SLOG_FATAL`
- `SLOG_PANIC`
- `SLOG_NONE`

File: README.md [hunk 3]
```diff
@@ -42,50 +49,100 @@ PRETTYLOG 0
...
+- `SLOG_FWARN`
+- `SLOG_FDEBUG`
+- `SLOG_FERROR`
+- `SLOG_FFATAL`
+- `SLOG_FPANIC`
+- `SLOG_FNONE`
 
 ### Print and log something
-Here is an example how use slog.
+Here is an example on how use slog.
```
Comment (Burnett01) [inline README.md:78]: "on how to use" :-)

File: src/slog.c [hunk 4]
```diff
@@ -71,347 +81,352 @@ void slog_get_date(SlogDate *sdate)
...
-    /* Check logging levels */
-    if(!level || level <= slg.level || level <= slg.file_level)
+
+    /* Check logging levels. */
+    if (!level || level <= slg.level || level <= slg.file_level)
     {
-        /* Handle flags */
-        switch(flag) 
+        /* Handle flags. */
+        switch (flag)
```
Comment (Burnett01) [inline src/slog.c:299]: In order to keep backward-compatibility, let's use this:

```c
/* Handle flags. */
switch (flag)
{
    case SLOG_LIVE:
    case SLOG_FLIVE:
        strncpy(color, CLR_NORMAL, sizeof(color));
        strncpy(alarm, "LIVE", sizeof(alarm));
        break;
    case SLOG_INFO:
    case SLOG_FINFO:
        strncpy(color, CLR_GREEN, sizeof(color));
        strncpy(alarm, "INFO", sizeof(alarm));
        break;
    case SLOG_WARN:
    case SLOG_FWARN:
        strncpy(color, CLR_YELLOW, sizeof(color));
        strncpy(alarm, "WARN", sizeof(alarm));
        break;
    case SLOG_DEBUG:
    case SLOG_FDEBUG:
        strncpy(color, CLR_BLUE, sizeof(color));
        strncpy(alarm, "DEBUG", sizeof(alarm));
        break;
    case SLOG_ERROR:
    case SLOG_FERROR:
        strncpy(color, CLR_RED, sizeof(color));
        strncpy(alarm, "ERROR", sizeof(alarm));
        break;
    case SLOG_FATAL:
    case SLOG_FFATAL:
        strncpy(color, CLR_RED, sizeof(color));
        strncpy(alarm, "FATAL", sizeof(alarm));
        break;
    case SLOG_PANIC:
    case SLOG_FPANIC:
        strncpy(color, CLR_RED, sizeof(color));
        strncpy(alarm, "PANIC", sizeof(alarm));
        break;
    case SLOG_NONE:
    case SLOG_FNONE:
    default:
        strncpy(prints, string, sizeof(string));
        flag = SLOG_FNONE;
        break;
}
```

**Review by Burnett01 — COMMENTED** [review]
Yet again - great work dude! In order to keep backward-compatibility, these changes should be taken into consideration.

File: src/slog.c (hunk 4 above)
Comment (Burnett01) [reply src/slog.c:299]: As you can see above, ``SLOG_NONE`` and ``SLOG_FNONE`` both behave like ``default`` so we don't need the code twice (Dry) and save an additional if-clause (line [341](https://github.com/kala13x/slog/pull/12/files/1ea113f80999d38e13b984744ac6c2a172dfeed7#diff-0b7de98ff3b11d118c9550e81a75f340R341) and [360](https://github.com/kala13x/slog/pull/12/files/1ea113f80999d38e13b984744ac6c2a172dfeed7#diff-0b7de98ff3b11d118c9550e81a75f340R360))

**Comment by GeorgeGkas** [comment]
Thank you very much @Burnett01 for your review. Backward compatibility is something that I have not consider. For the history part, I changed the flags because in the first test version of this update, I set the macro names the same as the flags (`SLOG_NONE`, `SLOG_FATAL`, etc), so I had to make this change. In the final revision I thought it was best to shift to lowercase names, but I forgot to change the flags back. I think we don't have to add more `case`s neither. Please, post review my last commit.

**Comment by Burnett01** [comment]
Great! Good work @GeorgeGkas

**Comment by Burnett01** [comment]
Hey @kala13x You up to get this merged? :) :+1:

**Comment by seachdamh** [comment]
@GeorgeGkas Great work.

I have also issue.

 -> I need use logging library completly without config file. Still program have to configure things like enable write to file not only at start but also - change parameters in "working time" (like lvl of logged messages). Can I make it with present version without forking / changing source code of library?

**Comment by GeorgeGkas** [comment]
@seachdamh This version I propose (1.5) keeps backward compatibility. If your are able to do this in the current stable release, then you are able to do this as well with 1.5. If there is no such option then you can propose an enchantment in issues page. For any further discussion, please open a new issue.

**Pull request merged**